osquery/config/packs.cpp

```diff
@@ -0,0 +1,184 @@
+/*
+ *  Copyright (c) 2014, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <boost/property_tree/json_parser.hpp>
+
+#include <osquery/core.h>
+#include <osquery/config/packs.h>
+#include <osquery/logger.h>
+#include <osquery/sql.h>
+
+
+namespace osquery {
+
+Pack::Pack(const std::string& name, const ptree& tree) {
+  initialize(name, "", tree);
+}
+
+Pack::Pack(const std::string& name,
+           const std::string& source,
+           const ptree& tree) {
+  initialize(name, source, tree);
+}
+
+Pack::Pack(const std::string& name, const std::string& json) {
+  std::stringstream stream;
+  stream << json;
+  ptree tree;
+  pt::read_json(stream, tree);
+  initialize(name, "", tree);
+}
+
+Pack::Pack(const std::string& name,
+           const std::string& source,
+           const std::string& json) {
+  std::stringstream stream;
+  stream << json;
+  ptree tree;
+  pt::read_json(stream, tree);
+  initialize(name, source, tree);
+}
+
+void Pack::initialize(const std::string& name,
+                      const std::string& source,
+                      const ptree& tree) {
+  name_ = name;
+  source_ = source;
+  discovery_queries_.clear();
+  if (tree.count("discovery") > 0) {
+    for (const auto& item : tree.get_child("discovery")) {
+      discovery_queries_.push_back(item.second.get_value<std::string>());
+    }
+  }
+
+  platform_.clear();
+  if (tree.count("platform") > 0) {
+    platform_ = tree.get<std::string>("platform");
+  }
+
+  version_.clear();
+  if (tree.count("version") > 0) {
+    version_ = tree.get<std::string>("version");
+  }
+
+  schedule_.clear();
+  if (tree.count("queries") > 0) {
+    for (const auto& q : tree.get_child("queries")) {
+      if (q.second.count("platform")) {
+        if (!checkPlatform(q.second.get<std::string>("platform"))) {
+          continue;
+        }
+      }
+
+      if (q.second.count("version")) {
+        if (!checkVersion(q.second.get<std::string>("version"))) {
+          continue;
+        }
+      }
+
+      ScheduledQuery query;
+      query.interval = q.second.get<int>("interval");
+      query.query = q.second.get<std::string>("query");
+      query.options["snapshot"] = q.second.get<bool>("snapshot", false);
+      query.options["removed"] = q.second.get<bool>("removed", true);
+      schedule_[q.first] = query;
+    }
+  }
+}
+
+std::map<std::string, ScheduledQuery> Pack::getSchedule() const {
+  return schedule_;
+}
+
+std::vector<std::string> Pack::getDiscoveryQueries() {
```

Comment:
Same as above? If so, would it be helpful to plebs like me if we had an inline comment to describe 1 use case, and how it's only safe from `Pack`'s perspective only to copy?